tests/managers/astprocd/test_usercode_lifecycle.py

```diff
@@ -0,0 +1,242 @@
+
+    def __init__(self) -> None:
+        self.times_called = 0
+        self.called_queue: List[CodeStatus] = []
+
+    def callback(self, status: CodeStatus) -> None:
+        """Mock inform callback."""
+        self.times_called += 1
+        self.called_queue.append(status)
+
+    @classmethod
+    def setup(
+        cls,
+        mount_path: Path = Path(),
+    ) -> Tuple[UsercodeLifecycle, 'StatusInformTestHelper']:
+        """Setup a lifecycle and helper for testing."""
+        sith = cls()
+        ucl = UsercodeLifecycle(
+            uuid=DiskUUID("foo"),
+            disk_info=DiskInfo(
+                uuid="foo",
+                mount_path=mount_path,
+                disk_type=DiskType.USERCODE,
+            ),
+            status_inform_callback=sith.callback,
+        )
+        return ucl, sith
+
+
+def test_usercode_lifecycle_init() -> None:
+    """Test the constructor."""
+    ucl, sith = StatusInformTestHelper.setup()
+    assert sith.times_called == 1
+    assert sith.called_queue == [CodeStatus.STARTING]
+
+    assert ucl._dir_path.exists()
+    assert ucl._process is None
+
+
+def test_usercode_lifecycle_properties() -> None:
+    """Test that the properties on the lifecycle work."""
+    ucl, _ = StatusInformTestHelper.setup()
+
+    assert ucl.uuid == "foo"
+    assert ucl.disk_info.uuid == "foo"
+    assert ucl.status is CodeStatus.STARTING
+
+
+def test_usercode_lifecycle_inform_callback_called_on_status_change() -> None:
+    """Test that the inform callback is called in the correct order."""
+    ucl, sith = StatusInformTestHelper.setup()
+
+    ucl.status = CodeStatus.RUNNING
+    ucl.status = CodeStatus.CRASHED
+    ucl.status = CodeStatus.RUNNING
+    ucl.status = CodeStatus.FINISHED
+
+    assert sith.called_queue == [
+        CodeStatus.STARTING,
+        CodeStatus.RUNNING,
+        CodeStatus.CRASHED,
+        CodeStatus.RUNNING,
+        CodeStatus.FINISHED,
+    ]
+
+
+def test_extract_and_validate_no_zip() -> None:
+    """Test that a disk with no zip is handled."""
+    ucl, sith = StatusInformTestHelper.setup(EXTRACT_ZIP_DATA / "no_zip")
+    assert not ucl._extract_and_validate_zip_file()
+
+    # Check that the log file contains the right text
+    log_file = EXTRACT_ZIP_DATA / "no_zip" / "log.txt"
+    with log_file.open("r") as fh:
+        assert fh.read() == "Unable to start code.\nUnable to find robot.zip file.\n"
+    log_file.unlink()  # clean up log file
+
+
+def test_bad_zip() -> None:
+    """Test that an invalid zip is handled."""
+    ucl, sith = StatusInformTestHelper.setup(EXTRACT_ZIP_DATA / "bad_zip")
+    assert not ucl._extract_and_validate_zip_file()
+
+    # Check that the log file contains the right text
+    log_file = EXTRACT_ZIP_DATA / "bad_zip" / "log.txt"
+    with log_file.open("r") as fh:
+        assert fh.read() == "Unable to start code.\nThe provided robot.zip is not a valid ZIP archive.\n"  # noqa: E501
+    log_file.unlink()  # clean up log file
```

Comment:
can this easily be made to happen whether or not the assertion fails? (maybe a try/finally around every single one, or if you can stick it in a teardown method or write a new context manager or something that might be nice)

Reply:
bdd22c7